File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/infras/persistences/jbdc_model/Account.kt
package dev.auguste.agni_api.infras.persistences.jbdc_model

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import dev.auguste.agni_api.core.entities.Account
import dev.auguste.agni_api.core.entities.Color
import dev.auguste.agni_api.core.entities.enums.AccountType
import dev.auguste.agni_api.core.entities.interfaces.IAccountDetail
import dev.auguste.agni_api.core.value_objects.BrokingAccountDetail
import dev.auguste.agni_api.core.value_objects.BusinessAccountDetail
import dev.auguste.agni_api.core.value_objects.CheckingAccountDetail
import dev.auguste.agni_api.core.value_objects.CreditCardAccountDetail
import dev.auguste.agni_api.core.value_objects.SavingAccountDetail
import dev.auguste.agni_api.infras.persistences.IMapper
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import org.springframework.stereotype.Component
import java.util.UUID

@Table("accounts")
data class JbdcAccountModel(
    @Id
    @get:JvmName("getIdentifier")
    @Column("account_id")
    val id: UUID,
    @Column("title")
    val name: String,
    val balance: Double,
    val type: String,
    @Column("currency_id")
    val currencyId: UUID?,
    val detail: String?,
    val color: String
) : JdbcModel() {
    override fun getId(): UUID {
        return id
    }
}

@Component
class JdbcAccountModelMapper(
    private val objectMapper: ObjectMapper
): IMapper<JbdcAccountModel, Account> {
    override fun toDomain(model: JbdcAccountModel): Account {
        var detail: IAccountDetail = CheckingAccountDetail(0.0)

        if (model.detail != null) {
            val detailMap = model.detail.let {
                objectMapper.readValue<Map<String, Any>?>(it)
            } ?: emptyMap()


            detail = when(AccountType.fromString(model.type)) {
                AccountType.SAVING -> SavingAccountDetail.fromMap(detailMap)
                AccountType.CREDIT_CARD-> CreditCardAccountDetail.fromMap(detailMap)
                AccountType.CHECKING -> CheckingAccountDetail.fromMap(detailMap)
                AccountType.BROKING-> BrokingAccountDetail.fromMap(detailMap)
                AccountType.BUSINESS -> BusinessAccountDetail(0.0)
            }
        }

        return Account(
            id = model.id,
            title = model.name,
            balance = model.balance,
            currencyId = model.currencyId,
            detail = detail,
            color = Color(model.color)
        )
    }

    override fun toModel(entity: Account): JbdcAccountModel {
        val detailJson = jacksonObjectMapper().writeValueAsString(entity.detail.toMap())

        return JbdcAccountModel(
            id = entity.id,
            name = entity.title,
            type = entity.detail.getType().value,
            balance = entity.balance,
            currencyId = entity.currencyId,
            detail = detailJson,
            color = entity.color.toString()
        )
    }

    override fun getEntityModelFieldName(): Map<String, String> {
        TODO("Not yet implemented")
    }

    override fun getTableName(): String {
        TODO("Not yet implemented")
    }

    override fun getSortField(): Set<String> {
        return setOf("title")
    }

    override fun getModelClass(): Class<JbdcAccountModel> {
        TODO("Not yet implemented")
    }
}

File: apps/agni-server_archived/src/infrastructure/persistences/models/account.ts
import { Account } from "@core/domains/entities/account";
import Mapper, { KnexTable } from "./mapper";
import { Knex } from "knex";
import { AccountType, mapperTypeAccount } from "@core/domains/constants";
import { KnexModel } from "./model";
import { BrockageAccountDetail } from "@core/domains/valueObjects/brockageAccount";
import { CreditCardAccountDetail } from "@core/domains/valueObjects/creditCardAccount";

export class KnexAccountTable implements KnexTable {
    getTableName(): string {
        return 'accounts'
    }

    async createTable(knex: Knex): Promise<void> {
        if (!(await knex.schema.hasTable('accounts')))
            await knex.schema.createTable("accounts", (table) => {
                table.uuid('account_id').primary()
                table.string('title')
                table.string('type')
                table.float('balance')
                table.string('currency_id')
                table.json('detail').nullable() 
                // table.float('credit_limit').defaultTo(0)
            });
    }
}

export type AccountModel = KnexModel & {
    account_id: string
    title: string
    type: string
    balance: number
    currency_id: string
    detail?: any
}

export class AccountModelMapper implements Mapper<Account, AccountModel> {
    toDomain(model: AccountModel): Account {
        let detail = undefined
        switch(model.type) {
            case AccountType.BROKING:
                detail = BrockageAccountDetail.fromJson(model.detail)
                break
            case AccountType.CREDIT_CARD:
                detail = CreditCardAccountDetail.fromJson(model.detail)
                break
            default:
                break
        }
        return new Account(
            model.account_id, 
            model.title, 
            mapperTypeAccount(model.type),
            model.currency_id,
            detail,
            model.balance,
        )
    }

    fromDomain(entity: Account): AccountModel {
        return {
            account_id: entity.getId(),
            title: entity.getTitle(),
            balance: entity.getBalance(),
            currency_id: entity.getCurrencyId(),
            detail: entity.getDetail()?.getJson(),
            type: entity.getType()
        }
    }

    getSortFilterFields(): string[] {
        return ['balance']
    }
    getIdField(): string {
        return 'account_id'
    }
    getNameField(): string {
        return 'title'
    }
} 
